expat/lib/xmlparse.c

```diff
@@ -2028,41 +2030,81 @@ XML_GetBuffer(XML_Parser parser, int len)
   default: ;
   }
 
-  if (len > parser->m_bufferLim - parser->m_bufferEnd) {
+  if (parser->m_bufferEnd) {
+    bufferAvailBytes = parser->m_bufferLim - parser->m_bufferEnd;
+  }
+  else {
+    bufferAvailBytes = 0;
+  }
+  if (len > bufferAvailBytes) {
 #ifdef XML_CONTEXT_BYTES
     int keep;
 #endif  /* defined XML_CONTEXT_BYTES */
-    /* Do not invoke signed arithmetic overflow: */
-    int neededSize = (int) ((unsigned)len + (unsigned)(parser->m_bufferEnd - parser->m_bufferPtr));
+    int neededSize;
+    int bufferTotalBytes;
+
+    if (parser->m_bufferPtr) {
+      /* Do not invoke signed arithmetic overflow: */
+      neededSize = (int) ((unsigned)len + (unsigned)(parser->m_bufferEnd - parser->m_bufferPtr));
+    }
+    else {
+      neededSize = len;
+    }
     if (neededSize < 0) {
       parser->m_errorCode = XML_ERROR_NO_MEMORY;
       return NULL;
     }
 #ifdef XML_CONTEXT_BYTES
-    keep = (int)(parser->m_bufferPtr - parser->m_buffer);
-    if (keep > XML_CONTEXT_BYTES)
-      keep = XML_CONTEXT_BYTES;
+    if (parser->m_bufferPtr) {
+      keep = (int)(parser->m_bufferPtr - parser->m_buffer);
+      if (keep > XML_CONTEXT_BYTES)
+        keep = XML_CONTEXT_BYTES;
+    }
+    else {
+      keep = 0;
+    }
     neededSize += keep;
 #endif  /* defined XML_CONTEXT_BYTES */
-    if (neededSize  <= parser->m_bufferLim - parser->m_buffer) {
+    if (parser->m_buffer) {
+      bufferTotalBytes = parser->m_bufferLim - parser->m_buffer;
+    }
+    else {
+      bufferTotalBytes = 0;
+    }
+    if (neededSize <= bufferTotalBytes) {
 #ifdef XML_CONTEXT_BYTES
-      if (keep < parser->m_bufferPtr - parser->m_buffer) {
-        int offset = (int)(parser->m_bufferPtr - parser->m_buffer) - keep;
+      int bufferParsedBytes;
+
+      if (parser->m_buffer) {
+        bufferParsedBytes = parser->m_bufferPtr - parser->m_buffer;
+      }
+      else {
+        bufferParsedBytes = 0;
+      }
+      if (keep < bufferParsedBytes) {
+        int offset = (int)bufferParsedBytes - keep;
```

Comment:
I think that `int` cast is redundant.

Reply:
Yes, that was just stupid typing on my part.  Sorry.